utils/lldp.go

```diff
@@ -0,0 +1,179 @@
+/***
+Copyright 2015 Cisco Systems Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package utils
+
+import (
+	"errors"
+	"os/exec"
+	"strings"
+
+	log "github.com/Sirupsen/logrus"
+)
+
+// Implement utilities to fetch lldp information aobut connecting neighbors.
+// The lldp information can be used in non cloud environment to automate
+// networking configuration on the physical devices
+
+// LLDPNeighbor lists the neighbor attributes as learned by the host
+type LLDPNeighbor struct {
+	SystemName        string
+	SystemDescription string
+	PortID            string
+	PortDesc          string
+	CiscoACIPodID     string
+	CiscoACINodeID    string
+}
+
+// FetchLLDPInfo reads the lldp information using lldptool
+// A native go lib could be done but it requires that lldpad daemon is
+// aceessible natively on the lldp ipc path
+func FetchLLDPInfo(ifname string) (*LLDPNeighbor, error) {
+	lldp := &LLDPNeighbor{}
+
+	output, err := exec.Command("docker", "exec", "lldpad", "lldptool",
+		"-n", "-t", "-i", ifname).CombinedOutput()
+	if err != nil {
+		log.Errorf("Error reading lldp information. Error: %s Output: \n%s\n", err, output)
+		return nil, err
+	}
+
+	err = parseLLDPArgs(string(output), lldp)
+
+	return lldp, err
+}
+
+func parseLLDPArgs(lldpToolOutput string, lldp *LLDPNeighbor) error {
+	var err error
+	lines := strings.Split(lldpToolOutput, "\n")
+	maxIdx := len(lines) - 1
+	for idx := 0; idx < maxIdx; {
+		line := strings.Trim(lines[idx], "\n")
+		idx++
+
+		if !strings.Contains(line, "TLV") ||
+			strings.Contains(line, "End of LLDPDU TLV") {
+			continue
+		}
+
+		attrIdx := idx
+		for !strings.Contains(lines[attrIdx], "TLV") && attrIdx <= maxIdx {
+			lines[attrIdx] = strings.Trim(lines[attrIdx], "\n")
+			attrIdx++
+		}
+		tlvLines := lines[idx:attrIdx]
+		idx = attrIdx
+		switch {
+		case strings.Contains(line, "System Name TLV"):
+			lldp.SystemName, err = lldpReadSystemName(tlvLines)
+		case strings.Contains(line, "System Description TLV"):
+			lldp.SystemDescription, err = lldpReadSystemDescription(tlvLines)
+		case strings.Contains(line, "Port ID TLV"):
+			lldp.PortID, err = lldpReadPortID(tlvLines)
+		case strings.Contains(line, "Port Description TLV"):
+			lldp.PortDesc, err = lldpReadPortDescription(tlvLines)
+		default:
```

Comment:
minor:
I think we don't need the `default` block if we initialize `err` to nil before `switch`